File: budget_core.js
const fs = require('fs');
const path = require('path');

const DATA_DIR = path.join(__dirname, 'data');
const DATA_FILE = path.join(DATA_DIR, 'budget.json');

const DEFAULT_CATEGORIES = {
  expense: ['Groceries', 'Clothes', 'Rent', 'Utilities', 'Transport', 'Entertainment', 'Health', 'Dining', 'Other'],
  earning: ['Salary', 'Freelance', 'Investments', 'Gifts', 'Other'],
};

const COLOR_PALETTE = [
  '#e74c3c','#3498db','#2ecc71','#f39c12','#9b59b6',
  '#1abc9c','#e67e22','#34495e','#e91e63','#c0392b',
  '#27ae60','#8e44ad','#f1c40f','#16a085','#d35400',
  '#2980b9','#7f8c8d','#2c3e50','#00bcd4','#ff5722',
];

function loadData() {
  try {
    return JSON.parse(fs.readFileSync(DATA_FILE, 'utf8'));
  } catch {
    return { entries: [], category_colors: {}, budgets: {} };
  }
}

function saveData(data) {
  fs.mkdirSync(DATA_DIR, { recursive: true });
  fs.writeFileSync(DATA_FILE, JSON.stringify(data, null, 2));
}

function ensureCategoryColor(data, category) {
  if (!data.category_colors) data.category_colors = {};
  if (!data.category_colors[category]) {
    const used = Object.values(data.category_colors);
    const color = COLOR_PALETTE.find(c => !used.includes(c)) || COLOR_PALETTE[Object.keys(data.category_colors).length % COLOR_PALETTE.length];
    data.category_colors[category] = color;
  }
  return data.category_colors[category];
}

function setBudget(category, limit, warningMargin) {
  const data = loadData();
  if (!data.budgets) data.budgets = {};
  data.budgets[category] = { limit, warning: warningMargin };
  saveData(data);
  return data.budgets[category];
}

function deleteBudget(category) {
  const data = loadData();
  if (!data.budgets || !data.budgets[category]) return false;
  delete data.budgets[category];
  saveData(data);
  return true;
}

function getBudgets() {
  const data = loadData();
  return data.budgets || {};
}

function checkBudget(category, newAmount, month, year) {
  const now = new Date();
  const m = (month || String(now.getMonth() + 1).padStart(2, '0'));
  const y = year || String(now.getFullYear());
  const data = loadData();
  const budget = (data.budgets || {})[category];
  if (!budget) return null;
  const currentTotal = (data.entries || [])
    .filter(e => e.type === 'expense' && e.category.toLowerCase() === category.toLowerCase() && e.date.startsWith(`${y}-${m}`))
    .reduce((s, e) => s + e.amount, 0);
  const newTotal = currentTotal + newAmount;
  const remaining = budget.limit - newTotal;
  let status = 'ok';
  if (newTotal > budget.limit) status = 'over';
  else if (remaining <= budget.warning) status = 'warning';
  return { status, spent: currentTotal, limit: budget.limit, warning: budget.warning, remaining, newTotal };
}

function addEntry(type, amount, category, description = '', date = null) {
  const data = loadData();
  ensureCategoryColor(data, category);
  const maxId = data.entries.reduce((m, e) => Math.max(m, e.id), 0);
  const entry = {
    id: maxId + 1,
    type,
    amount,
    category,
    description,
    date: date || new Date().toISOString().slice(0, 10),
  };
  data.entries.push(entry);
  saveData(data);
  return entry;
}

function deleteEntry(id) {
  const data = loadData();
  const before = data.entries.length;
  data.entries = data.entries.filter(e => e.id !== id);
  if (data.entries.length === before) return false;
  saveData(data);
  return true;
}

function listEntries(opts = {}) {
  const data = loadData();
  let entries = data.entries;
  if (opts.type) entries = entries.filter(e => e.type === opts.type);
  if (opts.category) entries = entries.filter(e => e.category.toLowerCase() === opts.category.toLowerCase());
  if (opts.month) entries = entries.filter(e => e.date.slice(5, 7) === opts.month.padStart(2, '0'));
  if (opts.year) entries = entries.filter(e => e.date.slice(0, 4) === opts.year);
  return entries.sort((a, b) => b.date.localeCompare(a.date));
}

function getSummary(opts = {}) {
  const entries = listEntries(opts);
  const summary = {};
  for (const e of entries) {
    summary[e.category] = (summary[e.category] || 0) + e.amount;
  }
  return summary;
}

function getMonthlyTotals(opts = {}) {
  const entries = listEntries(opts);
  const monthly = {};
  for (const e of entries) {
    const key = e.date.slice(0, 7);
    monthly[key] = (monthly[key] || 0) + e.amount;
  }
  return Object.fromEntries(Object.entries(monthly).sort());
}

function getCategories(type) {
  return DEFAULT_CATEGORIES[type] || [];
}

function getCategoryColors() {
  const data = loadData();
  return data.category_colors || {};
}

function getYearRange() {
  const data = loadData();
  const years = new Set(data.entries.map(e => e.date.slice(0, 4)));
  if (!years.size) return { min: String(new Date().getFullYear()), max: String(new Date().getFullYear()) };
  return { min: Math.min(...[...years].map(Number)).toString(), max: Math.max(...[...years].map(Number)).toString() };
}

function getYearMonths(year) {
  const data = loadData();
  const months = new Set(
    data.entries.filter(e => e.date.slice(0, 4) === year).map(e => e.date.slice(5, 7))
  );
  return [...months].sort();
}

module.exports = { addEntry, deleteEntry, listEntries, getSummary, getMonthlyTotals, getCategories, getCategoryColors, setBudget, deleteBudget, getBudgets, checkBudget, getYearRange, getYearMonths };

File: budget_core.py
import json
import os
from datetime import datetime
from collections import defaultdict

DATA_DIR = os.path.join(os.path.dirname(__file__), "data")
DATA_FILE = os.path.join(DATA_DIR, "budget.json")

COLOR_PALETTE = [
    "#e74c3c", "#3498db", "#2ecc71", "#f39c12", "#9b59b6",
    "#1abc9c", "#e67e22", "#34495e", "#e91e63", "#c0392b",
    "#27ae60", "#8e44ad", "#f1c40f", "#16a085", "#d35400",
    "#2980b9", "#7f8c8d", "#2c3e50", "#00bcd4", "#ff5722",
]

DEFAULT_CATEGORIES = {
    "expense": ["Groceries", "Clothes", "Rent", "Utilities", "Transport", "Entertainment", "Health", "Dining", "Other"],
    "earning": ["Salary", "Freelance", "Investments", "Gifts", "Other"]
}


def _load_data():
    if not os.path.exists(DATA_FILE):
        return {"entries": [], "category_colors": {}, "budgets": {}}
    with open(DATA_FILE, "r") as f:
        return json.load(f)


def _save_data(data):
    os.makedirs(DATA_DIR, exist_ok=True)
    with open(DATA_FILE, "w") as f:
        json.dump(data, f, indent=2)


def _ensure_category_color(data, category):
    if "category_colors" not in data:
        data["category_colors"] = {}
    if category not in data["category_colors"]:
        used = set(data["category_colors"].values())
        color = next((c for c in COLOR_PALETTE if c not in used), COLOR_PALETTE[len(data["category_colors"]) % len(COLOR_PALETTE)])
        data["category_colors"][category] = color
    return data["category_colors"][category]


def set_budget(category, limit, warning_margin):
    data = _load_data()
    if "budgets" not in data:
        data["budgets"] = {}
    data["budgets"][category] = {"limit": limit, "warning": warning_margin}
    _save_data(data)
    return data["budgets"][category]


def delete_budget(category):
    data = _load_data()
    if not data.get("budgets") or category not in data["budgets"]:
        return False
    del data["budgets"][category]
    _save_data(data)
    return True


def get_budgets():
    data = _load_data()
    return data.get("budgets", {})


def check_budget(category, new_amount, month=None, year=None):
    now = datetime.now()
    m = month or now.strftime("%m")
    y = year or now.strftime("%Y")
    data = _load_data()
    budget = data.get("budgets", {}).get(category)
    if not budget:
        return None
    current_total = sum(
        e["amount"] for e in data.get("entries", [])
        if e["type"] == "expense" and e["category"].lower() == category.lower() and e["date"].startswith(f"{y}-{m}")
    )
    new_total = current_total + new_amount
    remaining = budget["limit"] - new_total
    if new_total > budget["limit"]:
        status = "over"
    elif remaining <= budget["warning"]:
        status = "warning"
    else:
        status = "ok"
    return {"status": status, "spent": current_total, "limit": budget["limit"], "warning": budget["warning"], "remaining": remaining, "newTotal": new_total}


def add_entry(entry_type, amount, category, description="", date=None):
    data = _load_data()
    _ensure_category_color(data, category)
    entry = {
        "type": entry_type,
        "amount": amount,
        "category": category,
        "description": description,
        "date": date or datetime.now().strftime("%Y-%m-%d"),
        "id": len(data["entries"]) + 1 if data["entries"] else 1
    }
    if data["entries"]:
        entry["id"] = max(e["id"] for e in data["entries"]) + 1
    data["entries"].append(entry)
    _save_data(data)
    return entry


def delete_entry(entry_id):
    data = _load_data()
    before = len(data["entries"])
    data["entries"] = [e for e in data["entries"] if e["id"] != entry_id]
    if len(data["entries"]) == before:
        return False
    _save_data(data)
    return True


def list_entries(entry_type=None, category=None, month=None, year=None):
    data = _load_data()
    entries = data["entries"]
    if entry_type:
        entries = [e for e in entries if e["type"] == entry_type]
    if category:
        entries = [e for e in entries if e["category"].lower() == category.lower()]
    if month:
        entries = [e for e in entries if e["date"][5:7] == month.zfill(2)]
    if year:
        entries = [e for e in entries if e["date"][:4] == year]
    return sorted(entries, key=lambda e: e["date"], reverse=True)


def get_summary(entry_type=None, month=None, year=None):
    entries = list_entries(entry_type=entry_type, month=month, year=year)
    summary = defaultdict(float)
    for e in entries:
        summary[e["category"]] += e["amount"]
    return dict(summary)


def get_monthly_totals(entry_type=None, year=None):
    entries = list_entries(entry_type=entry_type, year=year)
    monthly = defaultdict(float)
    for e in entries:
        month_key = e["date"][:7]
        monthly[month_key] += e["amount"]
    return dict(sorted(monthly.items()))


def get_categories(entry_type):
    return DEFAULT_CATEGORIES.get(entry_type, [])


def get_category_colors():
    data = _load_data()
    return data.get("category_colors", {})


def get_year_range():
    data = _load_data()
    years = set(e["date"][:4] for e in data.get("entries", []))
    if not years:
        now = datetime.now()
        return {"min": now.strftime("%Y"), "max": now.strftime("%Y")}
    return {"min": min(years), "max": max(years)}


def get_year_months(year):
    data = _load_data()
    months = set(e["date"][5:7] for e in data.get("entries", []) if e["date"][:4] == year)
    return sorted(months)
